erizo/src/erizo/media/ExternalOutput.cpp

```diff
@@ -410,9 +413,14 @@ bool ExternalOutput::initContext() {
     if (context_->oformat->flags & AVFMT_GLOBALHEADER) {
       audio_stream_->codec->flags |= CODEC_FLAG_GLOBAL_HEADER;
     }
-
-    context_->streams[0] = video_stream_;
+    // TODO(nvazquezg): this 'if' is to avoid a matroska error when no video available: 'Invalid packet stream index: 1'
+    if (!init_video) {
```

Comment:
When we use init_video to set video stream to null, we don't have a proper recorded video.
The init_context is called more than once. If video_stream_ has been initialized before some audio data are being written, init_video stays as false and context->streams[0] is set to null stream which garbles the recording file.

As far as I understand, if video_codec_ is null, it means we won't receive video data.
Also same for audio. if audio_codec_ is null, there will be no audio data.

So instead of using init_video to check if this is just an audio stream, we can check video_codec_:

`

    //TODO: this 'if' is to avoid a matroska error when no video available: 'Invalid packet stream index: 1'
    if (video_codec_ == AV_CODEC_ID_NONE) {
        // To avoid the following matroska errors, we add CODEC_FLAG_GLOBAL_HEADER...
        // - Codec for stream 0 does not use global headers but container format requires global headers
        // - Only audio, video, and subtitles are supported for Matroska.
        video_stream_ = avformat_new_stream(context_, nullptr);
        video_stream_->codec->flags |= CODEC_FLAG_GLOBAL_HEADER;
        context_->streams[0] = video_stream_;
    }

`
That fixed my problem. Now, video is also being recorded properly.
Also, setting the global header is required to avoid some matrosko errors.

Reply:
After some tests, I've realized that recorded video is shorter than expected.
Using the master branch version (without this fix), the video is 5 seconds shorter (for a 15 second recording). I think that's because the remaining video frames in the queue cannot be recorded properly.

With this fix, 5 second part is missing at the start as well. I think that's because the audio data come after and re-writes the file. That's a guess, I'm not sure.

I couldn't get the recording feature work in a stable mode.

I think some work needs to be done about recording feature.

Reply:
Thanks for the feedback, I'll try to improve this behaviour and come back.